### PR TITLE
Update TD-SpeakerBeam

### PR DESCRIPTION
**Pull request by Emrys365**
This PR mainly update the implementation of TD-SpeakerBeam for target speaker extraction:

  1. It now also support the speaker embedding as an auxiliary input.
  2. A pre-mask activation is added by default to make the training easier.

### CODE REVIEW

**Comment by sw005320** [comment]
Can you add a result and model link to README.md?

**Comment by Emrys365** [comment]
OK. After I finish the training, I will do that.

File: egs2/librimix/tse1/conf/tuning/train_enh_tse_td_speakerbeam_4gpu_max.yaml [hunk 1]
```diff
@@ -1,9 +1,9 @@
 optim: adam
-max_epoch: 400
+max_epoch: 100
 batch_type: folded
 batch_size: 16
 iterator_type: chunk
-chunk_length: 24000
+chunk_length: 48000
```
Comment (simpleoier) [inline egs2/librimix/tse1/conf/tuning/train_enh_tse_td_speakerbeam_4gpu_max.yaml:6]: Do you think it may be useful to mention the sampling rate used for this parameter?

File: egs2/librimix/tse1/conf/tuning/train_enh_tse_td_speakerbeam_4gpu_max.yaml [hunk 2]
```diff
@@ -39,20 +39,20 @@ model_conf:
     share_encoder: true
 
 train_spk2enroll: data/train-100/spk2enroll.json
-enroll_segment: 24000
+enroll_segment: 48000
```
Comment (simpleoier) [inline egs2/librimix/tse1/conf/tuning/train_enh_tse_td_speakerbeam_4gpu_max.yaml:42]: ditto.

File: egs2/librimix/tse1/conf/tuning/train_enh_tse_td_speakerbeam_4gpu_max.yaml [hunk 3]
```diff
@@ -39,20 +39,20 @@ model_conf:
...
 load_spk_embedding: false
 load_all_speakers: false
 
 encoder: conv
 encoder_conf:
     channel: 256
-    kernel_size: 16
-    stride: 8
+    kernel_size: 32
```
Comment (simpleoier) [inline egs2/librimix/tse1/conf/tuning/train_enh_tse_td_speakerbeam_4gpu_max.yaml:49]: ditto.

File: espnet2/enh/extractor/td_speakerbeam_extractor.py [hunk 4]
```diff
@@ -118,10 +140,18 @@ def forward(
...
         feature = feature.transpose(1, 2)  # B, N, L
-        aux_feature = aux_feature.transpose(1, 2)  # B, N, L'
+        if self.use_spk_emb:
+            # B, N, L'=1
+            if input_aux.dim() == 2:
+                aux_feature = input_aux.unsqueeze(-1)
+            elif input_aux.size(-2) == 1:
+                aux_feature = input_aux.moveaxis(-2, -1)
+        else:
+            aux_feature = aux_feature.transpose(1, 2)  # B, N, L'
```
Comment (simpleoier) [inline espnet2/enh/extractor/td_speakerbeam_extractor.py:150]:
```suggestion
            assert aux_feature.dim() == 3
            aux_feature = aux_feature.transpose(1, 2)  # B, N, L'
```

File: espnet2/enh/extractor/td_speakerbeam_extractor.py [hunk 5]
```diff
@@ -118,10 +140,18 @@ def forward(
         B, L, N = feature.shape
 
         feature = feature.transpose(1, 2)  # B, N, L
-        aux_feature = aux_feature.transpose(1, 2)  # B, N, L'
+        if self.use_spk_emb:
+            # B, N, L'=1
+            if input_aux.dim() == 2:
```
Comment (simpleoier) [inline espnet2/enh/extractor/td_speakerbeam_extractor.py:145]: Is this expected to use `input_aux` instead of `aux_feature` here and after?

File: espnet2/enh/extractor/td_speakerbeam_extractor.py [hunk 6]
```diff
@@ -118,10 +140,18 @@ def forward(
         B, L, N = feature.shape
 
         feature = feature.transpose(1, 2)  # B, N, L
-        aux_feature = aux_feature.transpose(1, 2)  # B, N, L'
+        if self.use_spk_emb:
```
Comment (simpleoier) [inline espnet2/enh/extractor/td_speakerbeam_extractor.py:143]: It's a bit difficult to follow/understand in what cases `use_spk_emb=True` if I didn't know speakerbeam. Can you add some introduction comment here or above ?

File: espnet2/enh/layers/tcn.py [hunk 7]
```diff
@@ -94,9 +96,20 @@ def __init__(
         # [M, B, K] -> [M, C*N, K]
         mask_conv1x1 = nn.Conv1d(B, C * self.out_channel, 1, bias=False)
         # Put together (for compatibility with older versions)
-        self.network = nn.Sequential(
-            layer_norm, bottleneck_conv1x1, temporal_conv_net, mask_conv1x1
-        )
+        if pre_mask_nonlinear == "linear":
+            self.network = nn.Sequential(
```
Comment (simpleoier) [inline espnet2/enh/layers/tcn.py:100]: A minor question, if the `self.network` is used separately in forward (e.g. bottleneck, tcn, masknet called individually), what's the benefit of defining them in a Sequential(). I found it is a bit ambiguous in naming the subnet in forward().

**Review by simpleoier — COMMENTED** [review]
Thanks! I left some comments.

File: egs2/librimix/tse1/conf/tuning/train_enh_tse_td_speakerbeam_4gpu_max.yaml (hunk 1 above)
Comment (Emrys365) [reply egs2/librimix/tse1/conf/tuning/train_enh_tse_td_speakerbeam_4gpu_max.yaml:6]: Yes, this is true. I think I'd better mention the sample rate in the file name.

File: espnet2/enh/extractor/td_speakerbeam_extractor.py (hunk 5 above)
Comment (Emrys365) [reply espnet2/enh/extractor/td_speakerbeam_extractor.py:145]: Oh I think it should be `aux_feature`, although usually `input_aux` is equivalent here.

File: espnet2/enh/layers/tcn.py (hunk 7 above)
Comment (Emrys365) [reply espnet2/enh/layers/tcn.py:100]: This is just for **back-compatibility** with Conv-TasNet. Because previously the TCN implementation also used `Sequential` for speech separation, which can be used as a whole. But for TD-SpeakerBeam, we cannot use the `Sequential` module directly because of the input/output mismatch between sub-modules.

File: espnet2/enh/extractor/td_speakerbeam_extractor.py (hunk 6 above)
Comment (Emrys365) [reply espnet2/enh/extractor/td_speakerbeam_extractor.py:143]: I will add a comment above:
```python
# NOTE(wangyou): When `self.use_spk_emb` is True, `aux_feature` is assumed to be
# a speaker embedding; otherwise, it is assumed to be an enrollment audio.
if self.use_spk_emb:
...
```

**Review by simpleoier — APPROVED** [review]
LGTM!

**Comment by sw005320** [comment]
Hmm
https://github.com/espnet/espnet/actions/runs/4879274957/jobs/8705692129?pr=5155#step:8:35
Not sure. Should we wait for the fairseq to fix this?

**Comment by Emrys365** [comment]
> Hmm https://github.com/espnet/espnet/actions/runs/4879274957/jobs/8705692129?pr=5155#step:8:35 Not sure. Should we wait for the fairseq to fix this?

It seems this issue has been noticed since Jan 17, but they have not fixed it yet.

**Comment by Emrys365** [comment]
The model has also been uploaded to HuggingFace: https://huggingface.co/espnet/Wangyou_Zhang_librimix_train_enh_tse_td_speakerbeam_raw
